glib/src/byte_array.rs

```diff
@@ -187,27 +187,31 @@ impl fmt::Debug for ByteArray {
 
 macro_rules! impl_cmp {
     ($lhs:ty, $rhs: ty) => {
+        #[allow(clippy::redundant_slicing)]
         impl<'a, 'b> PartialEq<$rhs> for $lhs {
             #[inline]
             fn eq(&self, other: &$rhs) -> bool {
                 self[..].eq(&other[..])
```

Comment:
Why not fix this instead? You don't need the slicing here

Reply:
Lazyness mostly.

Reply:
Removing the `&` and `[..]` is less work than writing `#[allow(clippy::redundant_slicing)]` :rofl:

Reply:
Without it, we have two cases:

First by removing `[..]`:

```
error[E0277]: can't compare `byte_array::ByteArray` with `&&[u8]`
   --> glib/src/byte_array.rs:193:22
    |
193 |                 self.eq(&other)
    |                      ^^ no implementation for `byte_array::ByteArray == &&[u8]`
...
224 | impl_cmp!(ByteArray, &'a [u8]);
    | ------------------------------- in this macro invocation
    |
    = help: the trait `std::cmp::PartialEq<&&[u8]>` is not implemented for `byte_array::ByteArray`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: can't compare `byte_array::ByteArray` with `&std::vec::Vec<u8>`
   --> glib/src/byte_array.rs:193:22
    |
193 |                 self.eq(&other)
    |                      ^^ no implementation for `byte_array::ByteArray == &std::vec::Vec<u8>`
...
226 | impl_cmp!(ByteArray, Vec<u8>);
    | ------------------------------ in this macro invocation
    |
    = help: the trait `std::cmp::PartialEq<&std::vec::Vec<u8>>` is not implemented for `byte_array::ByteArray`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors
```

Second by removing the `&`:

```
error: function cannot return without recursing
   --> glib/src/byte_array.rs:192:13
    |
192 |             fn eq(&self, other: &$rhs) -> bool {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
193 |                 self.eq(other)
    |                 -------------- recursive call site
...
223 | impl_cmp!(ByteArray, [u8]);
    | --------------------------- in this macro invocation
    |
    = note: `-D unconditional-recursion` implied by `-D warnings`
    = help: a `loop` may express intention better if this is on purpose
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: function cannot return without recursing
   --> glib/src/byte_array.rs:192:13
    |
192 |             fn eq(&self, other: &$rhs) -> bool {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
193 |                 self.eq(other)
    |                 -------------- recursive call site
...
224 | impl_cmp!(ByteArray, &'a [u8]);
    | ------------------------------- in this macro invocation
    |
    = help: a `loop` may express intention better if this is on purpose
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: function cannot return without recursing
   --> glib/src/byte_array.rs:192:13
    |
192 |             fn eq(&self, other: &$rhs) -> bool {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
193 |                 self.eq(other)
    |                 -------------- recursive call site
...
225 | impl_cmp!(&'a ByteArray, [u8]);
    | ------------------------------- in this macro invocation
    |
    = help: a `loop` may express intention better if this is on purpose
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: function cannot return without recursing
   --> glib/src/byte_array.rs:192:13
    |
192 |             fn eq(&self, other: &$rhs) -> bool {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
193 |                 self.eq(other)
    |                 -------------- recursive call site
...
226 | impl_cmp!(ByteArray, Vec<u8>);
    | ------------------------------ in this macro invocation
    |
    = help: a `loop` may express intention better if this is on purpose
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: function cannot return without recursing
   --> glib/src/byte_array.rs:192:13
    |
192 |             fn eq(&self, other: &$rhs) -> bool {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
193 |                 self.eq(other)
    |                 -------------- recursive call site
...
227 | impl_cmp!(&'a ByteArray, Vec<u8>);
    | ---------------------------------- in this macro invocation
    |
    = help: a `loop` may express intention better if this is on purpose
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 5 previous errors
```

Reply:
`self.as_ref().eq(other)` is what you're looking for or `(&*self).eq(other)`

Reply:
If it's actually difficult and I'm just missing something, feel free to ignore me and keep it as it is :)

Reply:
It's actually one the strangest thing in rust: the auto-dereference isn't always working depending on the context. For example, you can't compare `&&T` with `&T` for some reason. Anyway, merging as is for the moment.

Reply:
Ok :)